### ISSUE TITLE
[FEATURE] Allow model parallelism option when training with multiple GPUs

### ISSUE THREAD

**Issue by tmostak** [h2oai/h2o-llmstudio#239]
### 🚀 Feature

Add way to specify model model parallelism when training using multiple GPUs.

### Motivation

Apologies if I'm getting this wrong but it appears currently when training models using multiple GPUs, the full model is replicated on each GPU (i.e. data parallelism). However, to train larger models, particularly for full fine-tuning (no LORA), allowing users to specify model parallelism where different layers of the model would be split among the available GPUs, would be helpful to prevent OOMs.

I just tried to do a full fine-tune of Open Llama 7B on a 8X40GB A100 instance but immediately hit an OOM (I could do a full-fine tune of the Open Llama 3B model on a single 80GB H100 with no issues).

Let me know if this feature already exists and I am just missing it, or if not, if it would be something you would consider as an enhancement. Thanks!

**Comment by psinger** [comment]
We started limited support for Pytorch 2.0 FSDP (this is what you mean with model parallelism), but lack stability and functionality there. LORA is in general hard to support with FSDP, and for full fine-tuning and also full training is not stable yet for FSDP.

We have it on the roadmap to support it better, but also depend on better FSDP support in Pytorch 2.0. Deepspeed is another option, but does not fit the current framework setup too well.

Would be great to have contributions here from the community.

That said, our recommend way over FSDP, is to use LORA and change the dtype. I am confident you will be able to fine tune open llama 7B on 40GB cards using `int4` as dtype.

![image](https://github.com/h2oai/h2o-llmstudio/assets/1677826/11553fdd-54c6-41cc-bd3e-4865518892b7)

**Comment by tmostak** [comment]
Thank you @psinger for the helpful clarification.

I was interested in trying full fine-tuning (vs LORA) after finding that with the Open Llama 3B model, full fine-tuning was resulting in relatively significant accuracy improvements compared to LORA (even using large 512-rank LORA).

Is FSDP worth trying in this case, or are you saying in the above that it's guaranteed not to work/be stable? (I unfortunately spun the big 8XA100 machine down, otherwise I'd just try).

**Comment by psinger** [comment]
My personal experience is contrary to what you observed. If you are mostly fine-tuning for stylistic changes on couple of thousand examples, I could not outperform LORA with full finetuning in any attempt I tried.

My personal recommendation is to stick to LORA, as it will also allow you to scale up the backbone size more easily given your hardware setups.

You can also check our cfg settings for open-llama-3b:
https://huggingface.co/h2oai/h2ogpt-gm-oasst1-en-2048-open-llama-3b/blob/main/cfg.yaml

I will dig into the full fine-tuning also again.

**Comment by tmostak** [comment]
@psinger I've generally read this as well, and seen papers showing LORA matching or beating the performance of full fine-tuning, but I think my use case, which is training the LLM on a set of ~7K text-to-SQL prompts in order to teach the model to generate correct SQL in response to user questions, might benefit from the much higher number of params (i.e. all of them) trained via a full fine-tune.

Just as one data point, the best perplexity on the validation set (taken from different SQL databases/tables than the training set) I've gotten with LORA using an Open Llama 3B backbone is 1.273, whereas full fine-tuning/pre-training yielded 1.254. Note that a LORA on the Open Llama 13B backbone yielded a perplexity of 1.2048, so my hope is that a full fine-tune on the 13B model would beat that still. Note that the perplexity differences seem small, but they translate into significant differences in BLEU score/real-world SQL generation performance, and for this use case, every bit of accuracy counts.

Unless you tell me it's guaranteed not to work, going to try FSDP tomorrow on the 7B and 13B models to see if I can get it working, and will report back with the results in case it's useful for others.

**Comment by judahkshitij** [comment]
@tmostak Did you end up trying FSDP feature in llmstudio on larger models like 7B and 13B models? How did it go? Were you able to fine tune these large models using FSDP strategy/feature or was it unstable and you could not use it? Please let us know how it went for you. Thank you.

**Comment by psinger** [comment]
There is currently ongoing work adding deepspeed functionality:
https://github.com/h2oai/h2o-llmstudio/pull/288

Happy for early feedback on that also @judahkshitij

**Comment by judahkshitij** [comment]
@psinger Why need to add Deepspeed if we have FSDP feature that we can test and make more stable? I thought deepspeed is low priority on the roadmap than making FSDP more stable.

Also, if we are supporting deepspeed, how can I use this feature and give early feedback? Just checkout deepspeed branch and the UI will show options to use deepspeed? 

Thanks for your message and thoughts.

**Comment by psinger** [comment]
We are trying out Deepspeed now instead of Pytorch 2.0 FSDP.

Exactly, checkout the branch and it should be available.

**Comment by judahkshitij** [comment]
@psinger I tried using deepspeed feature in LLM Studio and below are my observations and questions.

- I was able to train llama2-13b chat and base using LORA (values such as r=256, alpha=512) on 8 24gb GPUs, it takes time but the training finishes.

- I tried running Llama2-7b-base full finetuning(no LORA, with motivation same as @tmostak above that full finetuning will give better results than LORA) in fp16 with deepspeed on 8 GPUs with bs=1 and still got OOM error. Any thoughts on why? I thought with deepspeed I should be able to full finetune 7b model over 8 24gb GPUs.

- I tried Llama2-7b-base with large LORA r = 1024 (alpha=2048) (in hopes of making LORA competitive to full finetuning) in fp16 using deepspeed but training unstable(NaN in loss).

- As a test, I tried Llama2-7b-base with small LORA r = 32 (alpha=64) in fp16 using deepspeed but training unstable(NaN in loss). Is there an issue with using deepspeed with LORA when Llama2-7b-base model is involved? llama2-13b models trained well with LORA and deepspeed enabled as mentioned in my 1st bullet point.

- Aside: Also I tried Llama2-7b-base int8 and int4 full finetuning(no lora) without deepspeed but training unstable and loss becomes NaN. Any way to do full finetuning(no lora) in int8 and int4 or is it not supported yet?

Any thoughts on the above points are highly appreciated. Thanks.
